src/libsyntax/ext/tt/macro_rules.rs

```diff
@@ -274,3 +278,131 @@ pub fn compile<'cx>(cx: &'cx mut ExtCtxt,
+                // M.
+                let next_token = match tokens.peek() {
+                    // If T' closes a complex NT, replace T' with F
+                    Some(&&TtToken(_, CloseDelim(_))) => follow,
+                    Some(&&TtToken(_, ref tok)) => tok,
+                    // T' is any NT (this catches complex NTs, the next
+                    // iteration will die if it's a TtDelimited).
+                    Some(_) => continue,
+                    // else, we're at the end of the macro or sequence
+                    None => follow
+                };
+
+                // If T' is in the set FOLLOW(NT), continue. Else, reject.
+                match *next_token {
+                    Eof | MatchNt(..) => continue,
+                    _ if is_in_follow(cx, next_token, frag_spec.as_str()) => continue,
+                    ref tok => cx.span_err(sp, format!("`${0}:{1}` is followed by `{2}`, which \
+                                                        is not allowed for `{1}` fragments",
+                                                        name.as_str(), frag_spec.as_str(),
+                                                        token_to_string(tok))[])
+                }
+            },
+            TtSequence(_, ref seq) => {
+                // iii. Else, T is a complex NT.
+                match seq.separator {
+                    // If T has the form $(...)U+ or $(...)U* for some token U,
+                    // run the algorithm on the contents with F set to U. If it
+                    // accepts, continue, else, reject.
+                    Some(ref u) => check_matcher(cx, seq.tts[], u),
+                    // If T has the form $(...)+ or $(...)*, run the algorithm
+                    // on the contents with F set to EOF. If it accepts,
+                    // continue, else, reject.
+                    None => check_matcher(cx, seq.tts[], &Eof)
+                }
+            },
+            TtToken(..) => {
+                // i. If T is not an NT, continue.
+                continue
+            },
+            TtDelimited(_, ref tts) => {
+                // if we don't pass in that close delimiter, we'll incorrectly consider the matcher
+                // `{ $foo:ty }` as having a follow that isn't `}`
+                check_matcher(cx, tts.tts[], &tts.close_token())
+            }
+        }
+    }
+}
+
+fn is_in_follow(cx: &ExtCtxt, tok: &Token, frag: &str) -> bool {
+    if let &CloseDelim(_) = tok {
+        return true;
+    }
+
+    match frag {
+        "item" => {
+            // since items *must* be followed by either a `;` or a `}`, we can
+            // accept anything after them
+            true
+        },
+        "block" => {
+            // anything can follow block, the braces provide a easy boundary to
+            // maintain
+            true
+        },
+        "stmt" | "expr"  => {
+            match *tok {
+                Comma | Semi => true,
```

Comment:
Could we have at least one of FatArrow or RArrow here? Are we planning on using them in future expression syntax?